### PR TITLE
[libc++] Use availability to rely on key functions for bad_expected_access and bad_function_call

### PR DESCRIPTION
**Pull request by ldionne**
This patch uses our availability machinery to allow defining a key function for bad_function_call and bad_expected_access at all times but only rely on it when we can. This prevents compilers from complaining about weak vtables and reduces code bloat and the amount of work done by the dynamic linker.

rdar://111917845

### CODE REVIEW

**Comment by ldionne** [comment]
@philnik777 I remember you tried doing this in the past and ran into issues IIRC. Do you remember what those issues were?

File: libcxx/include/__expected/bad_expected_access.h [hunk 1]
```diff
@@ -27,10 +28,12 @@ _LIBCPP_BEGIN_NAMESPACE_STD
 template <class _Err>
 class bad_expected_access;
 
+#  if !_LIBCPP_AVAILABILITY_HAS_BAD_EXPECTED_ACCESS_KEY_FUNCTION
 _LIBCPP_DIAGNOSTIC_PUSH
 _LIBCPP_CLANG_DIAGNOSTIC_IGNORED("-Wweak-vtables")
+#  endif
```
Comment (philnik777) [inline libcxx/include/__expected/bad_expected_access.h:34]:
```suggestion
_LIBCPP_DIAGNOSTIC_PUSH
#  if !_LIBCPP_AVAILABILITY_HAS_BAD_EXPECTED_ACCESS_KEY_FUNCTION
_LIBCPP_CLANG_DIAGNOSTIC_IGNORED("-Wweak-vtables")
#  endif
```
Then we also don't have to guard the pop after the class.

File: libcxx/include/__functional/function.h [hunk 2]
```diff
@@ -54,8 +55,10 @@ _LIBCPP_BEGIN_NAMESPACE_STD
 
 // bad_function_call
 
+#  if !_LIBCPP_AVAILABILITY_HAS_BAD_FUNCTION_CALL_KEY_FUNCTION
```
Comment (philnik777) [inline libcxx/include/__functional/function.h:58]: Same suggestion here.

File: libcxx/include/__functional/function.h [hunk 3]
```diff
@@ -74,7 +77,9 @@ class _LIBCPP_EXPORTED_FROM_ABI bad_function_call : public exception {
   const char* what() const _NOEXCEPT override;
```
Comment (philnik777) [inline libcxx/include/__functional/function.h:77]: I think it would be good to add why we still consider this an ABI break, since it's not obvious.

**Review by philnik777 — APPROVED** [review]
LGTM with a green CI.

File: libcxx/include/__functional/function.h (hunk 3 above)
Comment (ldionne) [reply libcxx/include/__functional/function.h:77]: There was a comment in `__config`, I updated it and corrected it a little.

File: libcxx/src/expected.cpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <expected>
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+const char* bad_expected_access<void>::what() const noexcept { return "bad access to std::expected"; }
```
Comment (zibi2) [inline libcxx/src/expected.cpp:12]: @ldionne Louis, I noticed out down stream build failing with 
```
libcxx/src/expected.cpp:12:13: error: no template named 'bad_expected_access'
   12 | const char* bad_expected_access<void>::what() const noexcept { return "bad access to std::expected"; }
      |             ^
```
Shouldn't this definition be guarded with `#if _LIBCPP_STD_VER >= 23` to match its declaratoin in `libcxx/include/__expected/bad_expected_access.h` header file?

Comment (philnik777) [reply libcxx/src/expected.cpp:12]: The libc++ sources are always built with C++23. No need to guard stuff inside `src/`.

Comment (ldionne) [reply libcxx/src/expected.cpp:12]: @zibi2 What standard version are you compiling the sources with?

Comment (zibi2) [reply libcxx/src/expected.cpp:12]: > @zibi2 What standard version are you compiling the sources with?

For some reason we are compiling with -std=c++20 and thus the error.
Was there any recent change which started the -std=c++23 requirment?

Comment (ldionne) [reply libcxx/src/expected.cpp:12]: I don't remember when we did that, but I think it was necessary to implement a feature (potentially TZDB).

Comment (mordante) [reply libcxx/src/expected.cpp:12]: TZDB is C++20. `std::print` C++23, this required bumping the version. IIRC that was present LLVM 17.

Comment (zibi2) [reply libcxx/src/expected.cpp:12]: Thanks all. The `-std=c++20` was coming from down stream version of cmake which we need to update.

**Pull request merged**